login mutation

Sam09040 opened the pull request:
## add
- LoginInput interface
- login mutation on schema and resolver
- login tests
- jwt for authentication

Sam09040 committed:
Inicial setup

- test file with user mutation in testdb functioning

- add app folder and put interfaces, client and graphql content there

Sam09040 committed:
Small changes

- update prisma version
- correct index.test.mjs

Sam09040 committed:
small changes

- tsimp to run .ts
- run configuration with tsimp

- test file from .mjs to .ts
- resolvers import on index.ts

- .prettierrc.ts file as it was a duplicate

Sam09040 committed:
small change

- updated prisma import on resolvers file

Sam09040 committed:
error handling

- custom error to handle the errors
- error test file with four tests
- script to choose which test file to run

- error handling in resolvers
- test env file name

Sam09040 committed:
login mutation

- LoginInput interface
- login mutation on schema and resolver
- login tests
- jwt for authentication

Sam09040 committed:
remember me

- rememberMe on schema and LoginInput interface
- rememberMe logic with the token expiration time

-- tests to include remember me

Sam09040 committed:
authentication

- authentication for createUser mutation;
- authenticationError custom error;
- tests for authentication error;

Sam09040 committed:
small changes

- connection to the database and the server on the before all hook;

Sam09040 committed:
user query and tests

- authorization verification on user query
- user test file

- most of the logs

racsandoval commented on src/app/graphql/resolvers.ts:
```diff
@@ -124,6 +137,49 @@ export const resolvers = {
+
+      const user = await prisma.user.findUnique({ where: { email: email } });
+
+      if (!user) {
+        throw new CustomError('404', 'User not found!', {
+          field: 'email',
+          reason: 'The email you provided does not exist.',
+        });
+      }
+
+      const isValid = await comparePassword(password, user.password);
+      if (!isValid) {
+        throw new CustomError('400', 'Wrong password.', {
+          field: 'password',
+          reason: 'The provided password does not match.',
+        });
+      }
```
para autenticação, não é uma boa prática indicar que o email e senha estão incorretos, isso abre brechas pra um usuário malicioso ter informações dos usuários que existem no banco

Sam09040 replied:
então é melhor generalizar? falar que email ou senha está incorreto?

racsandoval commented on src/index.ts:
```diff
@@ -9,22 +9,31 @@ const server = new ApolloServer({
 
-server.listen(process.env.PORT).then(async ({ url }) => {
+/* server.listen(process.env.PORT).then(async ({ url }) => {
   console.log(url);
-});
+}); */
```
precisa desse comentário?

Sam09040 replied:
não, é algo que eu tenho consertado com o rebase e atualização das branches.

racsandoval commented on test/login.test.ts:
```diff
@@ -0,0 +1,221 @@
+  });
+
+  it('should return an error for invalid email', async () => {
+    const mutation = `
+            mutation login ($data: LoginInput!) {
+                login (data: $data) {
+                    user {
+                        id,
+                        name,
+                        email,
+                        birthDate
+                    },
+                    token
+                }
+            }
+        `;
```
essa variável se repete em todos os casos de teste, não daria pra extrair ela de cada `it`?

racsandoval commented on test/login.test.ts:
```diff
@@ -0,0 +1,221 @@
+    const login = response.data.data.login;
+
+    expect(response).to.have.property('status', 200);
+    expect(login).to.have.property('token').that.is.a('string');
```
poderia verificar se as propriedades do token estão corretas

racsandoval commented on test/login.test.ts:
```diff
@@ -0,0 +1,221 @@
+    expect(login.user.birthDate).to.equal('09-04-2004');
+  });
+
+  it('should return an error for missing token', async () => {
```
esse teste e o próximo não pertencem ao describe de criação de usuário?

Sam09040 replied:
para mim, eles faziam parte de testes relacionados ao login e existência de um token justamente por verificar se tem um token e ele é válido, além de não criar um usuário no final, por não testar o sucesso nesses casos, mas, realmente, facilita por todos usarem a mesma query.

racsandoval commented on test/user.test.ts:
```diff
@@ -0,0 +1,162 @@
+
+describe('user query', () => {
+    it('should return an error for no token', async () => {
+      const query = `
+              query user($userId: Int!){
+                user(id: $userId) {
+                  name,
+                  email,
+                  birthDate
+                }
+              }
+            `;
```
mesma coisa aqui

SamuelSsc commented on .gitignore:
```diff
@@ -3,3 +3,5 @@
 *.env*
 /prisma/migrations
+.vscode
+.tsimp
```
nit: linha vazia ao fim do arquivo

SamuelSsc commented on run-script.ts:
```diff
@@ -0,0 +1,12 @@
+import(`./${scriptFile}`).catch((err) => {
+    console.error("Error loading script: " + err);
+    process.exit(1);
+});
```
aqui tbm

Sam09040 committed:
corrections

- changed import for UserInput and LoginInput on resolvers.ts;
- email and password verification so it's not so specific;
- better organization of test files, making the query declared only once to be used by all test units;
- added jwt token verification on login success test;

racsandoval commented on test/error.test.ts:
```diff
@@ -0,0 +1,176 @@
+const url = `http://localhost:${port}/`;
+before(async () => {
+  try {
+    await server.listen(port).then(async ({ url }) => {
+      console.log(url);
+    });
```
aqui ao invés de utilizar `await` e `then` poderia colocar o console na linha abaixo ao `await`
```suggestion
    const { url } = await server.listen(port)
    console.log(url);
```

racsandoval commented on test/error.test.ts:
```diff
@@ -0,0 +1,176 @@
+
+describe('createUser mutation', () => {
+  it('should return an error for existing email', async () => {
+    const mutation = `
+            mutation createUser($data: UserInput!){
+                createUser(data: $data) {
+                    id,
+                    name,
+                    email,
+                    birthDate
+                }
+            }
+        `;
```
essa cont se repete 3x nesse arquivo

Sam09040 replied:
vai ser consertado com o rebase, perdão

racsandoval commented on test/login.test.ts:
```diff
@@ -0,0 +1,132 @@
+        `;
+
+  before(async () => {
+    try {
+      await server.listen(port).then(async ({ url }: any) => {
+        console.log(url);
+      });
+    } catch (error) {
+      console.log(error.message);
+    }
+
+    try {
+      await prisma.$connect();
+    } catch (error) {
+      console.log(error);
+    }
+    console.log(`listening on ${url}`);
+    console.log('connected to testdb');
```
como essas etapas se repetem em todos os testes, poderia criar uma função utilitária pra elas

racsandoval commented on test/login.test.ts:
```diff
@@ -0,0 +1,132 @@
+    expect(response).to.have.property('status', 200);
+    expect(login).to.have.property('token').that.is.a('string');
+    const isValid = jwt.verify(login.token, SECRET);
+    expect(isValid).to.have.property('userId').that.is.a('number');
```
além de verificar se é um número, poderia ver se é o número correto do id do usuário

sophia-bergamo reviewed (approved):
dica: pode clicar em `resolve conversation` em cada comentário que vc já resolveu, pra ficar melhor de visualizar o PR como um todo

Sam09040 committed:
better tests

## add:
- function to make before() and after() available for all tests, also having a possibility for user creation;
- object shorthand rule on eslint;
- expectations for expiration and id after token verification on login.test.ts;

## update:
- removed run-script.ts;

racsandoval commented on src/app/graphql/resolvers.ts:
```diff
@@ -1,8 +1,18 @@
+import { AuthenticationError } from 'apollo-server';
+const JWT_SECRET = process.env.JWT_SECRET ?? '';
+
+const comparePassword = async (password: string, hashedPassword: string) => {
+  const isMatch = await bcrypt.compare(password, hashedPassword);
+  return isMatch;
+};
```
poderia tipar o retorno da função
```suggestion
const comparePassword = async (password: string, hashedPassword: string): Promise<boolean> => {
  return bcrypt.compare(password, hashedPassword);
};
```

racsandoval commented on src/app/graphql/resolvers.ts:
```diff
@@ -19,9 +29,27 @@ export const resolvers = {
       return users;
     },
-    user: async (_: unknown, { id }: { id: number }) => {
+    user: async (_: unknown, { id }: { id: number }, { token }: any) => {
```
o `token` precisa ser `any`?

Sam09040 replied:
não

racsandoval commented on test/first.test.ts:
```diff
@@ -1,46 +1,15 @@
-  it('should create an user successfully', async () => {
-    const createUser = `
+describe('createUser mutation', () => {
+  StartFinish();
```
o que é `StartFinish`? o nome dessa função não deixa muito claro o objetivo dela

Sam09040 committed:
restructuring and typing

## add:
- ContextType to verify token typing and not use any;
- Promise<boolean> typing on comparePassword function;
- connect, disconnect and create util files for testing;

## update:
- corrected tests on error.test.ts;
